File: fitness-planner/engine/response_profiler.py
"""ResponseProfiler — 为后续宠物谱系积累训练响应，不做医学诊断。"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class ResponseObservation:
    completed_cycles: int
    adherence_pct: float
    performance_improvement_pct: float | None = None
    weight_trend_pct: float | None = None
    waist_trend_pct: float | None = None
    recovery_score: float | None = None


@dataclass
class ResponseProfile:
    maturity: str
    training_style: str
    metabolism_response: str
    muscle_gain_response: str
    fat_loss_response: str
    confidence: str

    def to_dict(self) -> dict:
        return self.__dict__.copy()


def profile_response(observation: ResponseObservation) -> ResponseProfile:
    if observation.adherence_pct >= 85:
        training_style = "steady"
    elif observation.adherence_pct >= 60:
        training_style = "variable"
    else:
        training_style = "still_learning"

    if observation.completed_cycles < 2:
        return ResponseProfile(
            maturity="preliminary",
            training_style=training_style,
            metabolism_response="unknown",
            muscle_gain_response="unknown",
            fat_loss_response="unknown",
            confidence="low",
        )

    performance = observation.performance_improvement_pct
    muscle = "unknown" if performance is None else ("responsive" if performance >= 5 else "steady")
    body_data = observation.weight_trend_pct is not None or observation.waist_trend_pct is not None
    fat = "unknown"
    if body_data:
        improved = (observation.weight_trend_pct or 0) < 0 or (observation.waist_trend_pct or 0) < 0
        fat = "responsive" if improved else "steady"

    # 代谢响应至少需要多周期身体趋势；这里只给训练响应标签，不输出医学快慢代谢。
    metabolism = "observed" if body_data else "unknown"
    known_axes = sum(value != "unknown" for value in (muscle, fat, metabolism))
    return ResponseProfile(
        maturity="developing" if observation.completed_cycles < 3 else "established",
        training_style=training_style,
        metabolism_response=metabolism,
        muscle_gain_response=muscle,
        fat_loss_response=fat,
        confidence="medium" if known_axes >= 2 else "low",
    )

File: flutter/lib/planner/response_profiler.dart
class ResponseObservation {
  const ResponseObservation({
    required this.completedCycles,
    required this.adherencePct,
    this.performanceImprovementPct,
    this.weightTrendPct,
    this.waistTrendPct,
    this.recoveryScore,
  });

  final int completedCycles;
  final double adherencePct;
  final double? performanceImprovementPct;
  final double? weightTrendPct;
  final double? waistTrendPct;
  final double? recoveryScore;
}

class ResponseProfile {
  const ResponseProfile({
    required this.maturity,
    required this.trainingStyle,
    required this.metabolismResponse,
    required this.muscleGainResponse,
    required this.fatLossResponse,
    required this.confidence,
  });

  final String maturity;
  final String trainingStyle;
  final String metabolismResponse;
  final String muscleGainResponse;
  final String fatLossResponse;
  final String confidence;

  Map<String, dynamic> toJson() => {
        'maturity': maturity,
        'training_style': trainingStyle,
        'metabolism_response': metabolismResponse,
        'muscle_gain_response': muscleGainResponse,
        'fat_loss_response': fatLossResponse,
        'confidence': confidence,
      };
}

ResponseProfile profileResponse(ResponseObservation observation) {
  final trainingStyle = observation.adherencePct >= 85
      ? 'steady'
      : (observation.adherencePct >= 60 ? 'variable' : 'still_learning');
  if (observation.completedCycles < 2) {
    return ResponseProfile(
      maturity: 'preliminary',
      trainingStyle: trainingStyle,
      metabolismResponse: 'unknown',
      muscleGainResponse: 'unknown',
      fatLossResponse: 'unknown',
      confidence: 'low',
    );
  }

  final performance = observation.performanceImprovementPct;
  final muscle = performance == null
      ? 'unknown'
      : (performance >= 5 ? 'responsive' : 'steady');
  final hasBodyData =
      observation.weightTrendPct != null || observation.waistTrendPct != null;
  final improved =
      (observation.weightTrendPct ?? 0) < 0 ||
      (observation.waistTrendPct ?? 0) < 0;
  final fat = hasBodyData ? (improved ? 'responsive' : 'steady') : 'unknown';
  final metabolism = hasBodyData ? 'observed' : 'unknown';
  final knownAxes = [
    muscle,
    fat,
    metabolism,
  ].where((value) => value != 'unknown').length;
  return ResponseProfile(
    maturity: observation.completedCycles < 3 ? 'developing' : 'established',
    trainingStyle: trainingStyle,
    metabolismResponse: metabolism,
    muscleGainResponse: muscle,
    fatLossResponse: fat,
    confidence: knownAxes >= 2 ? 'medium' : 'low',
  );
}
